hbase-client/src/main/java/org/apache/hadoop/hbase/client/HMasterAsyncRegistry.java

```diff
@@ -0,0 +1,236 @@
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProtos.GetMetaLocationsResponse;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProtos.IsActiveRequest;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProtos.IsActiveResponse;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProtos.MasterService;
+import org.apache.hbase.thirdparty.com.google.common.base.Preconditions;
+import org.apache.hbase.thirdparty.com.google.protobuf.ServiceException;
+
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Fetches the meta information directly from HMaster by making relevant RPCs. HMaster RPC end
+ * points are looked up via client configuration 'hbase.client.asyncregistry.masteraddrs' as
+ * comma separated list of host:port values. Should be set in conjunction with
+ * 'hbase.client.registry.impl' set to this class impl.
+ *
+ * The class does not cache anything. It is the responsibility of the callers to cache and
+ * avoid repeated requests.
+ */
+@InterfaceAudience.Private
+public class HMasterAsyncRegistry implements AsyncRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(HMasterAsyncRegistry.class);
+  public static final String CONF_KEY = "hbase.client.asyncregistry.masteraddrs";
+  private static final String DEFAULT_HOST_PORT = "localhost:" + HConstants.DEFAULT_MASTER_PORT;
+
+  // Parsed list of host and ports for masters from hbase-site.xml
+  private final List<ServerName> masterServers;
+  final Configuration conf;
+  // RPC client used to talk to the master servers. This uses a stand alone RpcClient instance
+  // because AsyncRegistry is created prior to creating a cluster Connection. The client is torn
+  // down in close().
+  final RpcClient rpcClient;
+  final int rpcTimeout;
+
+  public HMasterAsyncRegistry(Configuration config) {
+    masterServers = new ArrayList<>();
+    conf = config;
+    parseHortPorts();
+    // Passing the default cluster ID means that the token based authentication does not work for
+    // certain client implementations.
+    // TODO(bharathv): Figure out a way to fetch the CLUSTER ID using a non authenticated way.
+    rpcClient = RpcClientFactory.createClient(conf, HConstants.CLUSTER_ID_DEFAULT);
+    rpcTimeout = (int) Math.min(Integer.MAX_VALUE,
+        TimeUnit.MILLISECONDS.toNanos(conf.getLong(HBASE_RPC_TIMEOUT_KEY,
+            DEFAULT_HBASE_RPC_TIMEOUT)));
+  }
+
+  private void parseHortPorts() {
+    String hostPorts = conf.get(CONF_KEY, DEFAULT_HOST_PORT);
+    for (String hostPort: hostPorts.split(",")) {
+      masterServers.add(ServerName.valueOf(hostPort, ServerName.NON_STARTCODE));
+    }
+    Preconditions.checkArgument(!masterServers.isEmpty(), String.format("%s is empty", CONF_KEY));
+    // Randomize so that not every client sends requests in the same order.
+    Collections.shuffle(masterServers);
```

Comment:
is this shuffle pseudo random that will return same or diff result from different hosts?